lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs

```diff
@@ -0,0 +1,155 @@
+
+computeMinimumCoinForShelleyBasedEra
+    :: HasCallStack
+    => MinimumUTxOForShelleyBasedEra
+    -> TokenMap
+    -> Coin
+computeMinimumCoinForShelleyBasedEra
+    (MinimumUTxOForShelleyBasedEra era pp) tokenMap =
+        extractResult $
+        Cardano.calculateMinimumUTxO era
+            (embedTokenMapWithinPaddedTxOut era tokenMap)
+            (Cardano.fromLedgerPParams era pp)
+  where
+    extractResult :: Either Cardano.MinimumUTxOError Cardano.Value -> Coin
+    extractResult = \case
+        Right value ->
+            -- We assume that the returned value is a non-negative ada quantity
+            -- with no other assets. If this assumption is violated, we have no
+            -- way to continue, and must raise an error:
+            value
+                & unsafeValueToLovelace
+                & unsafeLovelaceToWalletCoin
+        Left e ->
+            -- We assume that the provided protocol parameters record has all
+            -- the required parameters for the given era. If this assumption is
+            -- violated, we have no way to continue, and must raise an error:
+            error $ unwords
+                [ "computeMinimumCoinForUTxO:"
+                , "unexpected error:"
+                , show e
+                ]
+
+-- | Embeds a 'TokenMap' within a padded 'Cardano.TxOut' value.
+--
+-- When computing the minimum UTxO quantity for a given 'TokenMap', we do not
+-- have access to an address or to an ada quantity.
+--
+-- However, in order to compute a minimum UTxO quantity through the Cardano
+-- API, we must supply a 'TxOut' value with a valid address and ada quantity.
+--
+-- It's imperative that we do not underestimate minimum UTxO quantities, as
+-- this may result in the creation of transactions that are unacceptable to
+-- the ledger. In the case of change generation, this would be particularly
+-- problematic, as change outputs are generated automatically, and users do
+-- not have direct control over the ada quantities generated.
+--
+-- However, while we cannot underestimate minimum UTxO quantities, we are at
+-- liberty to moderately overestimate them.
+--
+-- Since the minimum UTxO quantity function is monotonically increasing in the
+-- serialized length of its input, if we supply a 'TxOut' with an address and
+-- ada quantity whose serialized lengths are the maximum possible lengths, we
+-- can be confident that the resultant value will not be an underestimate.
+--
+embedTokenMapWithinPaddedTxOut
+    :: Cardano.ShelleyBasedEra era
+    -> TokenMap
+    -> Cardano.TxOut Cardano.CtxTx era
+embedTokenMapWithinPaddedTxOut era m =
+    toCardanoTxOut era $ TxOut dummyAddress $ TokenBundle dummyCoin m
+  where
+    dummyAddress :: Address
+    dummyAddress = Address $ BS.pack $ replicate maximumAddressLength 0
+      where
+        maximumAddressLength :: Int
+        maximumAddressLength = 57
+
+    dummyCoin :: Coin
+    dummyCoin = Coin $ intCast @Word64 @Natural $ maxBound
```

Comment:
Did any smaller values cause the property test to fail? 😃

Reply:
Also, I thought the address needed to be valid 🤔 This doesn't make the integration tests fail?

Reply:
> Did any smaller values cause the property test to fail? smiley (both re coin and addr)

For addresses, a shorter dummy address does indeed cause the property to fail:

```hs
  lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:90:9: 
  1) computeMinimumCoinForUTxO prop_computeMinimumCoinForUTxO_shelleyBasedEra_bounds
       Falsified (after 5 tests and 2 shrinks):
         TokenBundle {coin = Coin 0, tokens = TokenMap (fromList [])}
         ShelleyAddress Mainnet (KeyHashObj (KeyHash "799949ef8b7aad3c1753f658b190be1a919f5057f8ea8dd187b3a1de")) (StakeRefBase (KeyHashObj (KeyHash "dc3c5f13c7dec0337e4c53de6e675d3be50e6788d9d91ada26e10ffc")))
         ShelleyBasedEraBabbage ProtocolParameters {protocolParamProtocolVersion = (7,0), protocolParamDecentralization = Nothing, protocolParamExtraPraosEntropy = Nothing, protocolParamMaxBlockHeaderSize = 0, protocolParamMaxBlockBodySize = 0, protocolParamMaxTxSize = 2048, protocolParamTxFeeFixed = 0, protocolParamTxFeePerByte = 0, protocolParamMinUTxOValue = Nothing, protocolParamStakeAddressDeposit = Lovelace 0, protocolParamStakePoolDeposit = Lovelace 0, protocolParamMinPoolCost = Lovelace 0, protocolParamPoolRetireMaxEpoch = EpochNo 0, protocolParamStakePoolTargetNum = 100, protocolParamPoolPledgeInfluence = 0 % 1, protocolParamMonetaryExpansion = 0 % 1, protocolParamTreasuryCut = 0 % 1, protocolParamUTxOCostPerWord = Just (Lovelace 55999944), protocolParamCostModels = fromList [], protocolParamPrices = Just (ExecutionUnitPrices {priceExecutionSteps = 0 % 1, priceExecutionMemory = 0 % 1}), protocolParamMaxTxExUnits = Just (ExecutionUnits {executionSteps = 0, executionMemory = 0}), protocolParamMaxBlockExUnits = Just (ExecutionUnits {executionSteps = 0, executionMemory = 0}), protocolParamMaxValueSize = Just 0, protocolParamCollateralPercent = Just 150, protocolParamMaxCollateralInputs = Just 5}
         BS.length (unAddress maxLengthAddress)):
             29
         
         BS.length (unAddress (fromCardanoAddress address)):
             57
         
         BS.length (Cardano.serialiseToRawBytes address)):
             57
         
         ourResult:
             Coin 1420998579
         
         apiResultMaxBound:
             Coin 1420998579
         
         apiResultMinBound:
             Coin 1560998439
         
         Condition violated: ourResult >= apiResultMinBound

  To rerun use: --match "/computeMinimumCoinForUTxO/prop_computeMinimumCoinForUTxO_shelleyBasedEra_bounds/"
```

Reply:
> Also, I thought the address needed to be valid thinking This works?

Choosing an invalid address as the dummy address does indeed cause a **"malformed address"** err:
```hs
 Exception thrown while showing test case:
   INTERNAL ERROR: toCardanoTxOut: malformed address
   CallStack (from HasCallStack):
     error, called at lib/core/src/Cardano/Wallet/Util.hs:78:21 in main:Cardano.Wallet.Util
     internalError, called at lib/core/src/Cardano/Wallet/Util.hs:86:23 in main:Cardano.Wallet.Util
     tina, called at lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:1901:21 in main:Cardano.Wallet.Shelley.Compatibility
     toAlonzoTxOut, called at lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:1833:31 in main:Cardano.Wallet.Shelley.Compatibility
```

This apparently is a valid address:
```
maxLengthAddress :: Address
maxLengthAddress = Address $ BS.pack $ replicate maxAddressLength nullByte
  where
    maxAddressLength :: Int
    maxAddressLength = 56

    nullByte :: Word8
    nullByte = 0
```